File: matlab/classification/Dictionary_creation/create_reduced_dictionary.m
function feature_norm = create_reduced_dictionary(feature_training, N_sounds, database_folder, fold, type, data)

n_class = length(unique(feature_training(:,end-1)));

if length(N_sounds) == 1
    N_sounds = N_sounds * ones(1,n_class);
end

N_spect = size(feature_training,2)-2;

n_cluster = zeros(1,n_class);
idx_clusters = cell(1,n_class);

for k = 1:n_class
    feature_class = feature_training(feature_training(:,end-1) == k,:);
    n_cluster(k) = floor(size(feature_class,1) / N_sounds(k));
    % Clustering
    if strcmp(type,'single')
        if isempty(fold)
            %load(['Results/Dictionary reduction/',database_folder,'Clusters/Z_',num2str(k),'.mat'])
            load(['Results/Dictionary reduction/',database_folder,'Metaclasse/Z_',num2str(k),'.mat'])
        else
            %load(['Results/Mixture/Hierarchical/',database_folder,'Clusters/Fold ',num2str(fold),'/Z_',num2str(k),'.mat'])
            %load(['Results/Mixture/Hierarchical/',database_folder,'Metaclasse/Fold ',num2str(fold),'/Z_',num2str(k),'.mat'])
            load(['Clusters/',database_folder,'Metaclasse/Fold ',num2str(fold),'/Z_',num2str(k),'.mat'],'Z')
            %load(['Clusters/',database_folder,'Uniform/Fold ',num2str(fold),'/Z_',num2str(k),'.mat'],'Z')
        end
    elseif strcmp(type,'mixture')
        if isempty(fold)
            load(['Results/Dictionary reduction/',database_folder,'Mixtures/Z_',num2str(k),'.mat'])
        else
            load(['Results/Mixture/Hierarchical/',database_folder,'Mixtures/Fold ',num2str(fold),'/Z_',num2str(k),'.mat'])
        end
    end
    idx_clusters{k} = cluster(Z, 'maxclust', n_cluster(k));
    progressbar(k,n_class)
end

maxclust = sum(n_cluster);
my_feature = zeros(maxclust,N_spect+2);
cnt = 1;
for k = 1:n_class
    feature_class = feature_training(feature_training(:,end-1) == k,:);
    for nn = 1:n_cluster(k)
        my_feature(cnt,:) = [mean(feature_class(idx_clusters{k} == nn,1:end-2),1),k,1];
        cnt = cnt + 1;
    end
end

if strcmp(data,'raw')
    feature_norm = my_feature(:,1:end-2);
    feature_norm = [feature_norm, my_feature(:,end-1:end)];
elseif strcmp(data,'energy')
    feature_norm = my_feature(:,1:end-2);
elseif strcmp(data,'feature')
    feature_norm = my_feature(:,1:end-2) ./ repmat(sum(my_feature(:,1:end-2),2),[1,N_spect]);
    feature_norm = [feature_norm, my_feature(:,end-1:end)];
end

File: matlab/classification/Dictionary_creation/create_reduced_dictionary.py
import os
import numpy as np
import scipy.io as sio
from scipy.cluster.hierarchy import fcluster

def create_reduced_dictionary(feature_training, N_sounds, database_folder, fold, type_str, data_type):
    """
    Translated from create_reduced_dictionary.m
    """
    n_class = len(np.unique(feature_training[:, -2]))
    if np.isscalar(N_sounds) or len(np.atleast_1d(N_sounds)) == 1:
        N_sounds = np.ones(n_class) * np.atleast_1d(N_sounds)[0]
        
    N_spect = feature_training.shape[1] - 2
    n_cluster = np.zeros(n_class, dtype=int)
    idx_clusters = []
    
    for k in range(1, n_class + 1):
        feature_class = feature_training[feature_training[:, -2] == k, :]
        n_cluster[k-1] = int(np.floor(feature_class.shape[0] / N_sounds[k-1]))
        
        Z = None
        if type_str == 'single':
            if fold is None:
                path = f'Results/Dictionary reduction/{database_folder}Metaclasse/Z_{k}.mat'
            else:
                path = f'Clusters/{database_folder}Metaclasse/Fold {fold}/Z_{k}.mat'
            Z = sio.loadmat(path)['Z']
        elif type_str == 'mixture':
            if fold is None:
                path = f'Results/Dictionary reduction/{database_folder}Mixtures/Z_{k}.mat'
            else:
                path = f'Results/Mixture/Hierarchical/{database_folder}Mixtures/Fold {fold}/Z_{k}.mat'
            Z = sio.loadmat(path)['Z']
            
        idx = fcluster(Z, n_cluster[k-1], criterion='maxclust')
        idx_clusters.append(idx)
        print(f"Processed cluster {k}/{n_class}")
        
    maxclust = np.sum(n_cluster)
    my_feature = np.zeros((maxclust, N_spect + 2))
    cnt = 0
    for k in range(1, n_class + 1):
        feature_class = feature_training[feature_training[:, -2] == k, :]
        for nn in range(1, n_cluster[k-1] + 1):
            mean_feat = np.mean(feature_class[idx_clusters[k-1] == nn, :-2], axis=0)
            my_feature[cnt, :-2] = mean_feat
            my_feature[cnt, -2] = k
            my_feature[cnt, -1] = 1
            cnt += 1
            
    if data_type == 'raw':
        feature_norm = np.hstack([my_feature[:, :-2], my_feature[:, -2:]])
    elif data_type == 'energy':
        feature_norm = my_feature[:, :-2]
    elif data_type == 'feature':
        norm_factor = np.sum(my_feature[:, :-2], axis=1, keepdims=True)
        norm_factor[norm_factor == 0] = 1e-10
        feature_norm = np.hstack([my_feature[:, :-2] / norm_factor, my_feature[:, -2:]])
        
    return feature_norm
